tests/host_tools/change_config_space.c

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// This is used by the `test_config_space.py` integration test, which writes
+// into the microVM configured network device config space a new MAC address.
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+int show_usage() {
+    printf("Usage: ./change_net_config_space.bin [dev_addr_start] [number]\n");
+    printf("Example:\n");
+    printf("> ./change_net_config_space.bin 0xd00001000 0x060504030201\n");
+    return 0;
+}
+
+int main(int argc, char *argv[]) {
+    int fd, i, offset;
+    uint8_t *map_base, *virt_addr, mac_addr[18];
+
+    uint64_t mapped_size, page_size, offset_in_page, target;
+    uint64_t width = 6;
+
+    uint64_t config_offset = 0x100;
+    uint64_t device_start_addr = 0x00000000;
+    uint64_t mac = 0;
+
+    if (argc != 3) {
+        return show_usage();
+    }
+
+    device_start_addr = strtoull(argv[1], NULL, 0);
+    mac = strtoull(argv[2], NULL, 0);
+
+    fd = open("/dev/mem", O_RDWR | O_SYNC);
+    if (fd < 0) {
+        perror("Failed to open '/dev/mem'.");
+        return 1;
+    }
+
+    target = device_start_addr + config_offset;
+    // Get the page size.
+    mapped_size = page_size = getpagesize();
+    // Get the target address physical frame page offset.
+    offset_in_page = (unsigned) target & (page_size - 1);
+    /* If the data length goes out of the current page,
+     * double the needed map size. */
+    if (offset_in_page + width > page_size) {
+        /* This access spans pages.
+         * Must map two pages to make it possible. */
+        mapped_size *= 2;
+    }
+
+    // Map the `/dev/mem` to virtual memory.
+    map_base = mmap(NULL,
+            mapped_size,
+            PROT_READ | PROT_WRITE,
+            MAP_SHARED,
+            fd,
+            target & ~(off_t)(page_size - 1));
+    if (map_base == MAP_FAILED) {
+            perror("Failed to mmap '/dev/mem'.");
+            return 2;
+    }
+
+    // Write in the network device config space a new MAC.
+    virt_addr = (uint8_t*) map_base + offset_in_page;
+
+    *virt_addr = (uint8_t)(mac >> 40);
+    sprintf(mac_addr, "%02x", *virt_addr);
+    offset = 2;
+
+    *(virt_addr + 1) = (uint8_t) (mac >> 32);
+    sprintf(mac_addr + offset, ":%02x", *(virt_addr + 1));
+    offset += 3;
+
+    *(virt_addr + 2) = (uint8_t) (mac >> 24);
+    sprintf(mac_addr + offset, ":%02x", *(virt_addr + 2));
+    offset += 3;
+
+    *(virt_addr + 3) = (uint8_t) (mac >> 16);
+    sprintf(mac_addr + offset, ":%02x", *(virt_addr + 3));
+    offset += 3;
+
+    *(virt_addr + 4) = (uint8_t) (mac >> 8);
+    sprintf(mac_addr + offset, ":%02x", *(virt_addr + 4));
+    offset += 3;
+
+    *(virt_addr + 5) = (uint8_t) mac;
+    sprintf(mac_addr + offset, ":%02x", *(virt_addr + 5));
```

Comment:
nit: all of this can also be loop-ified

Reply:
Changing this to a `for` makes it fail on aarch64. It hangs and it is stopped due to integration test timeout > 300s.

Reply:
Writing to `/dev/mem` usually involves denoting the mapped memory as volatile, because it can be modified out of the context of the current thread. A summary for what happens is described bellow, based on [this](c0x.coding-guidelines.com/6.7.3.html):

"A volatile declaration may be used to describe an object corresponding to a memory-mapped input/output port or an object accessed by an asynchronously interrupting function. Actions on objects so declared shall not be ‘‘optimized out’’ by an implementation or reordered except as permitted by the rules for evaluating expressions."

Currently, this code hangs on `aarch64` because our integration tooling compiles it with `-O3`. Haven't looked into the generated asm on `aarch64` or `x86_64`, to understand how is the loop unrolled in both cases, but I took a shorter path and came out with the following plan:
- keep the implementation intent visible, marking the '/dev/mem' mapped memory as volatile.
- simplify the code, using the for loop.
- compile without any optimizations.

It seems that following this plan works well on `aarch64` too, it is intent clear, the code is reduced, as you suggested and we follow C spec recommendations. 

Moreover, following this plan on `aarch64` with `O3` compiler optimization, hangs as well, even if the compiler should omit optimizations around the volatile memory mapping (have to look into the generated asm to confirm or force skipping unrolling the loop).

Let me know what you think.

Reply:
Really cool deep dive, and really appreciate the thorough explanation. In this situation I'm OK with leaving the C code unchanged and the loop explicitly unrolled, maybe with a comment mentioning the compiler's behavior.

Reply:
Cool, OK. I'll keep the loop unrolled in the code and I'll add a comment related to loop unrolling behavior, when using `-O3`.